src/types/mod.rs

```diff
@@ -0,0 +1,811 @@
+    fn free(name: String) -> Self {
+        Constraint::Free(name)
+    }
+
+    fn unit() -> Self {
+        Constraint::Bound {
+            occurs_check: false,
+            bound: Bound::Unit,
+            concrete_data: Some(Arc::new(Final::unit())),
+        }
+    }
+
+    fn sum(a: Type, b: Type) -> Self {
+        Constraint::Bound {
+            occurs_check: false,
+            concrete_data: if let (Some(adata), Some(bdata)) = (a.final_data(), b.final_data()) {
+                Some(Arc::new(Final::sum(adata, bdata)))
+            } else {
+                None
+            },
+            bound: Bound::Sum(a, b),
+        }
+    }
+
+    fn product(a: Type, b: Type) -> Self {
+        Constraint::Bound {
+            occurs_check: false,
+            concrete_data: if let (Some(adata), Some(bdata)) = (a.final_data(), b.final_data()) {
+                Some(Arc::new(Final::product(adata, bdata)))
+            } else {
+                None
+            },
+            bound: Bound::Product(a, b),
+        }
+    }
+}
+
+impl Type {
+    /// Read-only pointer to data (used internally to make mutability easier to analyse)
+    fn inner_lock(&self) -> ReadOnlyMutexGuard {
+        ReadOnlyMutexGuard {
+            inner: self.inner.lock().unwrap(),
+        }
+    }
+
+    /// Clones the `Type`.
+    ///
+    /// This is the same as just calling `.clone()` but has a different name to
+    /// emphasize that what's being cloned is merely a ref-counted pointer.
+    pub fn shallow_clone(&self) -> Type {
+        self.clone()
+    }
+
+    /// Find the representative of this type variable in its set
+    ///
+    /// Used by the union-bound algorithm.
+    fn find_root(&self) -> Self {
+        let mut x = self.shallow_clone();
+        loop {
+            let x_lock = x.inner_lock();
+            // Find the parent of the current target, if there is one. If not, we
+            // are done, so return.
+            let parent = match x_lock.constraint {
+                Constraint::EqualTo { ref parent } => parent.shallow_clone(),
+                _ => {
+                    drop(x_lock);
+                    return x;
+                }
+            };
+            // If the parent has a parent, remove the intermediate link. (This is
+            // the "halving" variant of union-bound.)
+            let parent_lock = parent.inner_lock();
+            if let Constraint::EqualTo {
+                parent: ref grandparent,
+            } = parent_lock.constraint
+            {
+                let mut x_lock = x_lock.into_mutable(); // ok since x is `EqualTo` but only `Bound` constraints are finalized
+                x_lock.constraint = Constraint::EqualTo {
+                    parent: grandparent.shallow_clone(),
+                };
+                drop(x_lock);
+                x = grandparent.shallow_clone();
+            } else {
+                // If there is no grandparent, return the parent.
+                drop(parent_lock);
```

Comment:
Are these `drop`s necessary? We are not recursing. Maybe I am missing something

Reply:
Yes, for borrowck reasons. (Try removing the `drop` line and you'll get a compiler error.) I will add a comment about this.

Reply:
I guess you can't return an owned value or overwrite a value if there is a lock on it. I haven't worked with mutexes in a while